### ISSUE TITLE
Accept multiple matrices in `scipy.linalg.expm`

### ISSUE THREAD

**Issue by jond01** [scipy/scipy#12838]
**Is your feature request related to a problem? Please describe.**
I would like to compute exponentials of many matrices at once. I have a composite matrix of shape `(M, N, N)` and want the exponentials of the matrices in the last 2 dimensions (of shape `(N, N)`).
`scipy.linalg.expm` accepts only a single matrix, so a for loop or parallelized for loop can be used.

**Describe the solution you'd like**
Supporting multiple matrices as in [`numpy.linalg.eig`](https://numpy.org/doc/stable/reference/generated/numpy.linalg.eig.html#numpy.linalg.eig) for example.
Many functions in `numpy.linalg` support multiple matrices.

**Describe alternatives you've considered**
[tf.linalg.expm](https://www.tensorflow.org/api_docs/python/tf/linalg/expm)

**Additional context (e.g. screenshots)**
https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.expm.html

**Comment by DCtheTall** [comment]
@pvanmulbregt I'd be interested in writing a PR for this if it's available!

**Comment by rgommers** [comment]
This feature request makes sense in principle, but not in isolation. I think the real question is "do we want to support batches matrix operations in `scipy.linalg` functions. There are two reasons to allow this:

- Convenience - if you have a batch dimension, you typically want to use it many function calls, and writing a for-loop over the whole batch dimension can be a pain.
- Performance - significant performance gains are possible over a naive for-loop.

As gh-12946 shows, the 3-D version can simply be implemented as `np.asarray([linalg.expm(A_i) for A_i in A])`. This isn't any faster though. If we are going to implement such batched matrix operations, I think they should at least take care to figure out the right output shape, allocate that as a single contiguous 3-D array, and fill that in via the individual 2-D operations. Other optimizations may be possible.

Questions include:
- How many functions would this apply to?
- Should it be done for 3-D only, or also for higher-dimensional arrays? E.g. deep learning applications often have multiple batch dimensions, e.g. compare `torch.matmul/bmm`.
- How much work is this and do we think it's worth the effort?

**Comment by rgommers** [comment]
> Many functions in `numpy.linalg` support multiple matrices.

Yes indeed. With (I think) the unfortunate exception of the `dot`/`matmul`/`transpose` behaviour, which accept 3-D inputs but do not treat them as a batch of 2-D arrays.

**Comment by jond01** [comment]
Hi, first of all, thank you for the PR @DCtheTall!

@rgommers - great points. I admit that the issue is more profound.
Numpy indeed supports `matmul` and `dot` of a batch of 2-D arrays, `transpose` is also possible with the definition of axes.
I use it heavily in numpy, especially the matrix multiplication (`@`), and it flows harmonically with the broadcasting of other parts.

I agree that the implementation needs to be efficient to be worthwhile.
Regarding the questions:
- Ideally, all relevant functions (something like 50), realistically only the functions which do not have their counterpart in numpy (which is still a lot).
   Outside `linalg`, I know that [scipy.spatial.transform.Rotation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.transform.Rotation.html?highlight=multiple#scipy.spatial.transform.Rotation) supports a batch of rotations (only a single batch dimension is possible there). It is pretty convenient, hopefully also efficient.
- Also higher dimensional: 2d (no batch), 3d and more. Only the last 2 dimensions ("trailing dimensions") are the matrices.
- Honestly, I am not sure that it worth it when there are other implementations in _numpy_, _tensorflow_ of _pytorch_. However, when there is no other solution, I think it definitely does.

**Comment by ilayn** [comment]
Probably it would be better to make a small distinction between batch computation and stacked matrices (`ndarray`s with `n>=3`). For the latter there is a clear need. I am always concerned about that indeed and we have not been very active on this issue due to other issues (maintaining the 2D versions of codebase is already quite some undertaking). I have some preliminary work for solve, eig<family> and svd but far from being concrete. 

The other one is indeed a matter of convenience that we might not be bothered with since as mentioned we have to introduce all kinds of validations that they are all the same size etc.

**Comment by rgommers** [comment]
> Probably it would be better to make a small distinction between batch computation and stacked matrices (`ndarray`s with `n>=3`).

Could you explain the difference between the two terms? They sound like synonyms to me.

**Comment by jond01** [comment]
If I understand correctly @ilayn, the cases/inputs are:
1. batch computation: a list of 2d arrays, possibly with different sizes - `[(M1, M1), (M2, M2), ..., (Mn, Mn)]`
2. stacked matrices: a single array with 2 dimensions or more: `(..., M, M)`, the matrices reside in the last two indices

And we want to focus on 2

**Comment by rgommers** [comment]
> batch computation: a list of 2d arrays, possibly with different sizes

**Comment by DCtheTall** [comment]
Hey all, I was on vacation for the last week. I am back now and ready to start iterating on my PR.

> As gh-12946 shows, the 3-D version can simply be implemented as np.asarray([linalg.expm(A_i) for A_i in A])... I think they should at least take care to figure out the right output shape, allocate that as a single contiguous 3-D array, and fill that in via the individual 2-D operations. Other optimizations may be possible.

Pardon my naive implementation, I was hoping to get the ball rolling and I think that I was thinking this was for "stacked" matrices as well. I agree that allocating the higher-dimensional array first makes sense.

> I meant batch as in batch dimensions of deep learning applications.

Just so we're clear, the idea is to have linalg.expm support higher-dimensional arrays whose last 2 dimensions contain matrices. E.g.

```python
A = ... # Some numpy array with shape (M, N, d, d)
print(scipy.linalg.expm(A).shape) # (M, N, d, d)
```

If this is not correct, perhaps I need some clarification 😅

**Comment by rgommers** [comment]
> Just so we're clear, the idea is to have linalg.expm support higher-dimensional arrays whose last 2 dimensions contain matrices.

Yes indeed, although if we should support arbitrary dimensions or just 3-D would still be a question. Just to be clear, let's use the same dimension letters as the `np.matmul` docstring, which says: `(n,k), (k,m)->(n,m)`.

```
# So this would be: (b1, b2, n,k), (k,m)->(b1, b2, n,m)
>>> x = np.arange(5*24).reshape(2, 3, 4, 5)
>>> y = np.ones((5, 6))
>>> (x @ y).shape
(2, 3, 4, 6)
```

The question is how much work this would be to do for all functions in `linalg` where the idea applies. For `expm` it's obviously easy, but for something like `svd` it may get complicated quickly.

**Comment by jond01** [comment]
I am not sure why 3d+ (i.e. 3d, 4d, or more) is more difficult than 3d only.
If we have a 3d+ array `a`, we can transform it to 3d with:
```python
a_reshaped = a.reshape((-1, a.shape[-2], a.shape[-1]))
```
Then operate on `a_reshaped` with function for 3d arrays and use the original shape `a.shape` to have back the batch dimensions.
The small complications I see are: keeping the correct matrices order in reshaping, reshaping back in the correct way.

**Comment by ilayn** [comment]
It's not more difficult but rather current 2D versions have lots of consistency checks in place for early exits. nD versions needs to dismantle those and put new ones in place. Memory handling is slightly different in the nD case too especially if big arrays are expected (solve, eig etc.). So we can't just use any expensive linear algebra ops behind the scenes because this is supposed to bring in some benefits and not choking and slowdowns. So some nontrivial effort is needed especially on the testing suite side.

**Comment by DCtheTall** [comment]
@rgommers @ilayn thanks for the input!

> Yes indeed, although if we should support arbitrary dimensions or just 3-D would still be a question.
Any reason for `expm` to not support higher dimensions? I think doing nD is more flexible.

> So some nontrivial effort is needed especially on the testing suite side.
Noted. For `expm` are there existing perf tests? Not sure how you guys typically go about performance benchmarking.

**Comment by rgommers** [comment]
There's a `benchmarks` directory in the root of the repo. We use Airspeed Velocity (`asv`) to run benchmarks. There's an `expm` benchmark in `benchmarks/sparse_linalg_expm.py`.

Benchmarks get published at https://pv.github.io/scipy-bench/, so we have a change to detect regressions.

**Comment by DCtheTall** [comment]
@rgommers thanks, I'll update my PR sometime this coming week 😄

**Comment by DCtheTall** [comment]
> There's a benchmarks directory in the root of the repo. We use Airspeed Velocity (asv) to run benchmarks. There's an expm benchmark in benchmarks/sparse_linalg_expm.py.

Thanks, got that working 😄 

> I think they should at least take care to figure out the right output shape, allocate that as a single contiguous 3-D array, and fill that in via the individual 2-D operations. Other optimizations may be possible.

@rgommers in the case of `expm` the function as written is a pure function which initializes and returns a new array. If we want to allocate a higher dimensional array and do the 2D operations in its elements, it seems `expm` would need to be called in place. That seems like a bit more of a refactoring effort, but something I would be willing to try out if you think that's the right way to go about it.

**Comment by rgommers** [comment]
> If we want to allocate a higher dimensional array and do the 2D operations in its elements, it seems `expm` would need to be called in place. That seems like a bit more of a refactoring effort, but something I would be willing to try out if you think that's the right way to go about it.

A full in-place version seems like a lot of effort indeed, and is going to be a bit ugly because there's `return` statements all over the code and the implementation is split into many separate functions. So perhaps not worth it in this case. Would you agree @ilayn?

**Comment by ilayn** [comment]
Indeed, it is either a whole overhaul touching many places or we might not attempt it with a bandaid.

**Comment by DCtheTall** [comment]
@rgommers @ilayn thanks for the valuable input. [12946](https://github.com/scipy/scipy/pull/12946) is closed.

**Comment by ilayn** [comment]
@DCtheTall would you be willing to reconsider when I have some time to dig into this? We are fully aware that this section of the codebase needs much love but we don't have extra hands to do so. I might actually have some time end of February so if you want we can perform the necessary surgeries.

**Comment by DCtheTall** [comment]
> would you be willing to reconsider when I have some time to dig into this?... I might actually have some time end of February so if you want we can perform the necessary surgeries.

Sure, I'd be happy to lend a helping hand when the time comes. I'll keep an eye out for when you're free.

**Comment by DCtheTall** [comment]
@ilayn friendly ping

**Comment by ilayn** [comment]
Since @nschloe also raised a request for this, I guess we should start wearing the overalls

**Comment by ilayn** [comment]
By the way, inspired by the tensorflow implementation I'm going to attempt to get https://epubs.siam.org/doi/abs/10.1137/18M1228876 working instead of the sparse/dense amalgam which we currently have. That should reduce number of copies we are making to regularize the datatype. But talk is cheap so let's see. Fingers crossed

**Comment by ilayn** [comment]
Hmm for that you should use `lsim` from signal but I guess it doesn't support complex matrices. OK so it's still in the dynamics evolution context.

**Comment by nschloe** [comment]
> Just out of curiosity, since this is a very particular request it is very unlikely that you folks are asking this coincidentally,

I think this is a coincidence actually. I need exp for many small 3x3 matrices; this comes out of an ODE context, too.

**Comment by ilayn** [comment]
Some update on this: last evening I've checked more closely Fasi & Higham paper but the paper is specifically geared towards higher precision instead of lower which is the case for numpy/scipy so probably better to clean up the existing one it seems.

**Comment by ilayn** [comment]
Just to give some update about this, first my apologies for the ridiculous delays. I have written this function and some internals more than 5 times now and it is getting a bit crazy. After cythonizing and whatnot I ran out of brainz to explain how the middle section of the following plot (with the excellent `perfplot` of @nschloe ) come to be

![image](https://user-images.githubusercontent.com/1303842/140738638-83c8c51f-cb70-42d7-b61f-4d46823e1781.png)
![image](https://user-images.githubusercontent.com/1303842/140739574-f2ef3ae4-2287-49a9-ab79-9e353f2c3157.png)


Indeed up to n=1500 I can generate some speedup on average and then matrix multiplications start to dominate and we are back to old performance, but I cannot get rid of the #%@★❢ problematic section in the middle around 65-90. I can't explain that for the love of me but probably some L1, L2 cache thing happening but why it hits as bad as the SciPy version and why it is almost constant, I have no idea. But if I can spare more time, I'll look into it more. Any feedback is more than welcome.

PS: The individual `%timeit` in IPython actually shows a bigger performance gap between two algos but still the same issue around 65-90.

**Comment by ilayn** [comment]
Some more update (mostly for myself 😃 so I remember what I was doing previous weekend) here is the current situation; 

![image](https://user-images.githubusercontent.com/1303842/141592899-8a61757f-5c95-471c-a281-a87ed20a6029.png)

Many things are optimized now and finally the performance plots are sufficiently separated. There were two unaccounted issues. One of them was very expensive matrix transposes created tons of cache-misses and the second was, as it turns out that weird problem was coming from LAPACK routine `?getrf` which we are discussing at https://github.com/xianyi/OpenBLAS/issues/3448 hence if you can run this very small script and paste the result here that would be really beneficial for us. Especially if any MKL user is reading this that would be great so we know if it is OpenBLAS specific or not.

```python
import numpy as np
import scipy.linalg as la

for n in range(50, 120, 5):
    zzz = np.random.rand(n, n)
    %timeit la.lu_factor(zzz)
```

**Comment by nschloe** [comment]
>  if you can run this very small script and paste the result here that would be really beneficial for us

Ubuntu 21.10, Intel i5-8350U, SciPy 1.8.0.dev0+1902.eff1c87:
```python
111 µs ± 15 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
119 µs ± 17.5 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
131 µs ± 14.5 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
148 µs ± 16.1 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
260 µs ± 155 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
194 µs ± 25.7 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
211 µs ± 22.9 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
260 µs ± 127 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
191 µs ± 6.42 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
226 µs ± 36.6 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
310 µs ± 99.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
286 µs ± 60.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
317 µs ± 139 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
317 µs ± 75.8 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

**Comment by ilayn** [comment]
Thanks Nico is that an MKL (usually the case with conda) or OpenBLAS ?

**Comment by nschloe** [comment]
This is on a pip-installed scipy so I'm assuming OpenBLAS.

**Comment by rgommers** [comment]
Note that conda-forge is the best option for benchmarking here, because you can switch out all BLAS implementations without touching your `numpy` package installed in your environment: https://conda-forge.org/docs/maintainer/knowledge_base.html?highlight=blas#switching-blas-implementation

**Comment by ilayn** [comment]
I took Ralf's suggestion and installed conda with MKL. It turns out that I am quite stupid and while under the impression that it was the code's fault for the jumps above, I accidentally optimized way too much 🤦  

![image](https://user-images.githubusercontent.com/1303842/141665696-d595242c-b7f0-4374-b486-632f70f9e311.png)
![image](https://user-images.githubusercontent.com/1303842/141666009-abc446eb-8635-4322-9fd0-c6b253c78dad.png)

That's conda MKL and I haven't managed conda to switch to openblas. However that's not the point anyways. So there goes A LOT of time 😅 Anyways, I'll roll this out to other dtypes and then to stacked matrices then hopefully a PR. Thanks everyone for the patience.

**Comment by jond01** [comment]
> if you can run this very small script and paste the result here that would be really beneficial for us.

- macOS Monterey (12.0.1) (`macOS-10.16-x86_64-i386-64bit`)
- Processor: 2.3 GHz 8-Core Intel Core i9
- Memory: 32 GB 2667 MHz DDR4

With  `scipy` from `conda-forge` (MKL, python `3.9.5`, `scipy` version `1.7.1`):
```
24.1 µs ± 359 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
26.3 µs ± 273 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
28.3 µs ± 592 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
32.2 µs ± 720 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
34.4 µs ± 460 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
41.1 µs ± 856 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
41.6 µs ± 876 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
46.4 µs ± 1.29 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
50.2 µs ± 580 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
54.9 µs ± 773 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
58.5 µs ± 290 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
65.5 µs ± 1.01 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
70.2 µs ± 722 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
78.3 µs ± 1.43 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

With  `scipy` from `pip` (OpenBLAS, python `3.9.6`, `scipy` version `1.7.2`):
```
26 µs ± 255 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
30 µs ± 291 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
36.6 µs ± 3.77 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
39.5 µs ± 618 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
42.7 µs ± 1.44 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
49.4 µs ± 674 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
50.1 µs ± 825 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
58.9 µs ± 394 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
61.8 µs ± 900 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
74.3 µs ± 2.55 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
99.4 µs ± 6.64 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
102 µs ± 1.59 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
119 µs ± 3.55 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
132 µs ± 3.66 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

**Comment by ilayn** [comment]
I forgot to ping the kind folks on this issue for the PR should they wish to test it for themselves 😉 It's linked above. Any feedback welcome.

**Comment by jungerm2** [comment]
I don't mean to re-open this issue, but I was wondering if something like this could be implemented for `logm` too. It seems that it is not batched yet and that this would be a natural extension to this.

**Comment by ilayn** [comment]
My roadmap was to replace all linalg functions on the public API but then I got some build problems that I couldn't fix and my daily work made things even more difficult to spare time. 

Currently what I have is most of the `linalg.sqrtm` done (which `linalg.logm` uses underneath) and half of `linalg.solve` (#12824) so in case anyone else wants to jump in between please do so until I find some time.

**Comment by ilia10000** [comment]
Hey @ilayn! Just wanted to check if you had managed to get this working for `linalg.sqrtm`?

**Comment by mgbukov** [comment]
I see this was fixed a while ago; any reason why this feature was not added to logm, cosm, sinm, etc., as well?

**Comment by ilia10000** [comment]
Hey @ilayn! Just wanted to check back in on this and see if there were any updates for `sqrtm`?

**Comment by ilayn** [comment]
Hi @ilia10000 The batch input is already implemented but the rewrite is going to be v1.16; see #22406
